text/0000-naming-conventions.md

```diff
@@ -0,0 +1,267 @@
+- Start Date: 2018-03-21
+- RFC PR: (leave this empty)
+- WordPress Coding Standards Issue: (leave this empty)
+
+# Summary
+
+This guide is intended to maximize consistency when writing code in the WordPress environment.
+Naming your language constructs is important and should reflect the nature of what it does.
+
+# Basic example
+
+Naming variables, functions, classes, namespaces should be unified across the core, themes and plugins.
+
+Variables should be clear and understandable, written in lowercase using hyphens as a separator.
+
+```php
+$post_array;
+```
+
+Classes (namespaces, traits, interfaces, abstract classes and classes) should be written in capitalized words separated by underscores.
+
+```php
+namespace Theme_Name\Subnamespace;
+
+class Class_Name implements Interface {
+  // class implementation.
+}
+```
+
+PHP constants should be written in lowercase, while user defined constants should be written in uppercase separated by underscores.
+
+```php
+define( 'DOING_AJAX', false );
+```
+
+File names should be written using lowercase letters separated by hyphens, and class file names must be based on the class name, with the underscores in the class name replaced by hyphens.
+
+```php
+content-gallery.php
+
+class-name.php
+```
+
+# Motivation
+
+The motivation behind this is future proofing the WordPress code. Also the current status of WordPress core is a mess. There are mixed OOP constructs with functional programming, all in files that have various namings. At some point in the future PHP 5.2 will be deprecated in the WordPres core. This will enable a gradual, but necessary, rewrite of the WordPress core using the proper OOP principles. With that upgrade come all the modern PHP language constructs (namespaces, interfaces, autoloading etc.). The sooner we have an agreement around the proper naming, the better we'll be prepared to handle the possible rewrite. And it will help with the current [core autoloader proposal](https://core.trac.wordpress.org/ticket/36335).
+
+Once the naming proposal is accepted and unified, a gradual core rewrite can begin that follows the unified rule.
+
+The inspiration comes from:
+
+https://www.alainschlesser.com/interface-naming-conventions/
+https://www.php-fig.org/psr/psr-2/
+
+# Detailed design
+
+#### Keywords and constants
+
+PHP [keywords](http://php.net/manual/en/reserved.keywords.php) must be in lower case, as well as PHP constants `true`, `false` and `null`. User defined constants must be in all upper-case with underscores separating words
+
+```php
+<?php
+define( 'ENV', 'staging' );
+```
+
+#### Namespace and use declarations
+
+Namespace must be defined at the top of the file. There must be one blank line after the `namespace` declaration. Namespaces must follow the class names naming - capitalized words separated by underscores.
```

Comment:
I thought this proposal was about naming conventions ? So why does this section mention formatting guidelines ?

Reply:
> Namespaces must follow the class names naming

That sentence doesn't make sense and could be read as "the namespace should be the same as the class name", which I don't think is your intention.